circuit_passes/src/passes/const_arg_deduplication.rs

```diff
@@ -1,23 +1,27 @@
 use std::cell::RefCell;
-use std::collections::{HashMap, HashSet};
+use std::collections::HashMap;
 use code_producers::llvm_elements::stdlib::GENERATED_FN_PREFIX;
 use compiler::circuit_design::function::{FunctionCode, FunctionCodeInfo};
 use compiler::circuit_design::template::TemplateCode;
 use compiler::compiler_interface::Circuit;
 use compiler::hir::very_concrete_program::Param;
-use compiler::intermediate_representation::{InstructionPointer, BucketId, UpdateId, new_id};
+use compiler::intermediate_representation::{new_id, BucketId, InstructionPointer, UpdateId};
 use compiler::intermediate_representation::ir_interface::*;
 use compiler::intermediate_representation::translate::ARRAY_PARAM_STORES;
 use crate::bucket_interpreter::error::BadInterp;
 use crate::bucket_interpreter::memory::PassMemory;
 use crate::{default__name, default__get_mem};
 use super::{CircuitTransformationPass, GlobalPassData, builders};
 
+/// Extracts the constant argument initialization section of a template into
+/// *.array.param.* functions and replaces the original section with a call to
+/// the function. This reduces the overall size of the LLVM IR by allowing
+/// templates with the same constant arguments to reuse the same function.
+/// NOTE: See `post_hook_circuit` below because it defines another transformation
+/// that happens in this pass that is unrelated to constant argument deduplication.
```

Comment:
I think partly because it's so minor I couldn't justify cloning the entire AST again and partly because I'm lazy. But it should probably be separated into its own pass.

Reply:
Fair tbh. We should probably try to make the transformation passes zero-copy and move what we don't touch. But like lowest priority. I can create tickets for it if you agree

Reply:
I agree. I had the same thoughts when I was working on the performance but it will be a pretty big task.